ar71xx: add support for TP-LINK CPE510 V2.0

apcameron opened the pull request:
Adds Support for the TP-LINK CPE510 V2.0 by TP-Link.
The hardware is almost the same as the CPE510 V1.0
Follow the same processes as for the CPE510 V1.0

Signed-off-by: Andrew Cameron <apcameron@softhome.net>

apcameron commented:
No It should not as i have a cpe210v2 as well but using the earlier pull request that 937 was based on.

apcameron commented:
I have installed my code with pr #642 and tested it but not with #937

apcameron commented:
In looking at PR #937 it changes the logic for all devices that call tplink_pharos_board_detect
in target/linux/ar71xx/base-files/lib/ar71xx.sh and tplink_pharos_check_image in ar71xx/base-files/lib/upgrade/platform.sh

As a result of this all those devices should be tested as part of that Pull Request.

As a result of that my PR will need updating if PR #937 is implemented as is.

ecsv commented:
The CPE210 support was merged. Your PR must be rebased

apcameron commented:
I have updated my code to work with the code from PR #937 and pushed a new commit

ecsv commented on target/linux/ar71xx/base-files/lib/ar71xx.sh:
```diff
@@ -581,8 +581,18 @@ ar71xx_board_detect() {
+                        local part
+                        part=$(find_mtd_part 'product-info')
+                        dd if=$part bs=1 skip=4360 count=64 2>/dev/null | tr -d '\r\0' | head -n 1
+                        }
```
This (and the following lines) look extremely misaligned

ecsv commented on target/linux/ar71xx/base-files/lib/ar71xx.sh:
```diff
@@ -581,8 +581,18 @@ ar71xx_board_detect() {
+                        }
+                        md_string="$(tplink_v2_check)"
+                        if [ "${md_string:35:3}" == "2.0" ] || [ "${md_string:26:3}" == "2.0" ] ; then
+                                                tplink_pharos_board_detect "$(tplink_pharos_v2_get_model_string)"
```
Looks like you had to much tabs in storage which you had to spend somewhere.

adrianschmutzler replied:
You can just use tplink_pharos_board_detect "$md_string" here.

adrianschmutzler commented on target/linux/ar71xx/base-files/lib/ar71xx.sh:
```diff
@@ -581,8 +581,18 @@ ar71xx_board_detect() {
-		name="cpe510"
-		tplink_pharos_board_detect "$(tplink_pharos_get_model_string | tr -d '\r')"
+                name="cpe510"
+                tplink_v2_check() {
```
I do not really get the reason for this function inside the switch/case block; in addition, it is repeated identically in platform.sh. By the way, this is also identical to the tplink_pharos_v2_get_model_string() introduced for the CPE210v2. So you could just remove the function and use md_string="$(tplink_pharos_v2_get_model_string)" later.

adrianschmutzler commented on target/linux/ar71xx/base-files/lib/ar71xx.sh:
```diff
@@ -581,8 +581,18 @@ ar71xx_board_detect() {
+                        dd if=$part bs=1 skip=4360 count=64 2>/dev/null | tr -d '\r\0' | head -n 1
+                        }
+                        md_string="$(tplink_v2_check)"
+                        if [ "${md_string:35:3}" == "2.0" ] || [ "${md_string:26:3}" == "2.0" ] ; then
```
Does this (and the required tplink_pharos_v2_get_model_string ) work with a CPE510v1?

adrianschmutzler commented on target/linux/ar71xx/base-files/lib/upgrade/platform.sh:
```diff
@@ -589,6 +588,21 @@ platform_check_image() {
+                        }
+                        md_string="$(tplink_v2_check)"
+                        if [ "${md_string:35:3}" == "2.0" ] || [ "${md_string:26:3}" == "2.0" ] ; then
+				tplink_pharos_check_image "$1" "7f454c46" "$(tplink_pharos_v2_get_model_string)" '\0\xff\r' && return 0
```
Like above:
- Inline function declaration
- CPE510v1 compatibility
- reuse md_string

adrianschmutzler replied:
Sure about the "7f454c46" in contrast to the "01000000" on the CPE210v2?

adrianschmutzler commented:
If I'm not mistaken, you try to implement the cpe510 v1 and v2 under the same device definition (in contrast to splitting it into cpe510-v1 and cpe510-v2). This looks odd to me.
However, I only start to understand the structure of OpenWRT device implementations; maybe this is okay I am too pedantic. Advice would be helpful here ...

apcameron committed:
ar71xx: add support for TP-LINK CPE510 V2.0

Adds Support for the TP-LINK CPE510 V2.0 by TP-Link.
The hardware is almost the same as the CPE510 V1.0
Follow the same processes as for the CPE510 V1.0

Signed-off-by: Andrew Cameron <apcameron@softhome.net>

apcameron commented:
@ecsv I have tried to reformat the code. It looks ok when I look at the files.
@adrianschmutzler I have re-written the code to try to address your concerns.  I do not have a CPE510v1.0 so I could not test with that but I have tested with my CPE510V2.0 and have also tested that sysupgrade works. 
The reason I have not split it into cpe510v1 and cpe510-v2 is the hardware is almost the same Unlike the CPE210 V2.0 which was different.

adrianschmutzler commented on target/linux/ar71xx/base-files/lib/ar71xx.sh:
```diff
@@ -582,7 +582,15 @@ ar71xx_board_detect() {
-		tplink_pharos_board_detect "$(tplink_pharos_get_model_string | tr -d '\r')"
+		tplink_pharos_board_detect "$(tplink_pharos_v2_get_model_string)"
+		case $AR71XX_MODEL in
+                'TP-Link CPE510 v2.0')
```
Here you still have spaces instead of tabs.

adrianschmutzler commented on target/linux/ar71xx/base-files/lib/ar71xx.sh:
```diff
@@ -582,7 +582,15 @@ ar71xx_board_detect() {
 	*"CPE510/520")
 		name="cpe510"
-		tplink_pharos_board_detect "$(tplink_pharos_get_model_string | tr -d '\r')"
+		tplink_pharos_board_detect "$(tplink_pharos_v2_get_model_string)"
```
Thanks for tidying up; much clearer now.
However, I do not like the multiple occasions of tplink_pharos_board_detect. I would prefer to do the evaluation based on the model string, so save "$(tplink_pharos_v2_get_model_string)" into a variable:
modelstr="$(tplink_pharos_v2_get_model_string)"
and then do some case $modelstr in 'something' instead of your case $AR71XX_MODEL.
I don't thick its a good idea to use non-local variables for this local switch-case.
The same would be applicable for platform.sh.
However, I'm not a committer, so others may see it differently.

apcameron commented:
@adrianschmutzler I have further refined the code taking into account your comments.
However the value in $AR71XX_MODEL is set by tplink_pharos_board_detect and is the value I need
for the case statement. Looking elsewhere I see other uses of $AR71XX_MODEL in the code so mine 
is not the only time its referenced.

adrianschmutzler commented:
Okay for me. Don't have a CPE510 to test.
So now you will have to wait for testers and/or people with commit access to proceed.

hmh commented:
We do have a CPE510 v1.x here, but no v2.  Is it worth it to bother testing this pull request on the CPE 510 v1.1?

adrianschmutzler commented:
Can't judge, I only helped here because I was involved with the changes concerning the tplink_pharos_* functions for the CPE210v2; I cannot tell too certain whether the mixed v1/v2 for the CPE510 will work eventually.
However, testing should be relatively easy and the result will help us in any case. If something goes wrong, TFTP should be easy, too (at least it is for the CPE210v1.1).

apcameron commented:
It work hurt to test on a CPE510 v1.1 to make sure it does not interfere with the code for that device  however we need more people to test on a CPE510 v2.0
It is working on my CPE510 v2.0 without any issues and I have had someone test a CPE510v1.0 to make sure that still works.

blogic commented:
is testing finished ? the patch looks correct but i dont have access to any of the devices so i cant test myself.

hmh commented:
The CPE510 v1.x was on loan, and we should get it back today.  If I still can't manage to test before this weekend, please don't delay the PR merge on my regard.

Sorry about this delay.

apcameron commented:
I have it running on my TP-LINK CPE510 V2.0 without any problems.

blogic commented:
merged, Thanks !
